File: go/unicode.go
package gencmu

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

// unicodeTable is grammars/unicode.txt: the mark and alpha ranges that give
// a character its class (engine §1), and the simple lowercase mappings that
// lowercase() applies (engine §10). Every library reads this table rather
// than its platform's Unicode data, so that all of them agree.
type unicodeTable struct {
	marks, alphas [][2]rune
	lower         map[rune]rune
}

func parseUnicodeTable(text string) (*unicodeTable, error) {
	t := &unicodeTable{lower: map[rune]rune{}}
	for n, line := range strings.Split(text, "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		f := strings.Fields(line)
		hex := func(s string) (rune, error) {
			v, err := strconv.ParseUint(s, 16, 32)
			return rune(v), err
		}
		switch f[0] {
		case "unicode":
			continue
		case "mark", "alpha", "lower":
			if len(f) != 3 {
				return nil, fmt.Errorf("unicode.txt line %d: expected three fields", n+1)
			}
			a, err1 := hex(f[1])
			b, err2 := hex(f[2])
			if err1 != nil || err2 != nil {
				return nil, fmt.Errorf("unicode.txt line %d: bad code point", n+1)
			}
			switch f[0] {
			case "mark":
				t.marks = append(t.marks, [2]rune{a, b})
			case "alpha":
				t.alphas = append(t.alphas, [2]rune{a, b})
			default:
				t.lower[a] = b
			}
		default:
			return nil, fmt.Errorf("unicode.txt line %d: unknown entry %q", n+1, f[0])
		}
	}
	for _, r := range [][][2]rune{t.marks, t.alphas} {
		sort.Slice(r, func(i, j int) bool { return r[i][0] < r[j][0] })
	}
	return t, nil
}

func inRanges(ranges [][2]rune, c rune) bool {
	i := sort.Search(len(ranges), func(i int) bool { return ranges[i][1] >= c })
	return i < len(ranges) && ranges[i][0] <= c
}

// class is the weak class tag of a character token (engine §1).
func (t *unicodeTable) class(c rune) string {
	switch {
	case c >= 0x09 && c <= 0x0D, c == 0x20, c == 0x85, c == 0xA0, c == 0x1680,
		c >= 0x2000 && c <= 0x200A, c == 0x2028, c == 0x2029, c == 0x202F, c == 0x205F, c == 0x3000:
		return "space"
	case c >= '0' && c <= '9':
		return "digit"
	case inRanges(t.marks, c):
		return "mark"
	case inRanges(t.alphas, c):
		return "alpha"
	}
	return "other"
}

func (t *unicodeTable) lowercase(s string) string {
	var b strings.Builder
	for _, c := range s {
		if l, ok := t.lower[c]; ok {
			c = l
		}
		b.WriteRune(c)
	}
	return b.String()
}

// fnv1a64 is the hash that keys precompiled DOMs (engine §8).
func fnv1a64(text string) string {
	h := uint64(0xcbf29ce484222325)
	for i := 0; i < len(text); i++ {
		h ^= uint64(text[i])
		h *= 0x100000001b3
	}
	s := strconv.FormatUint(h, 16)
	return strings.Repeat("0", 16-len(s)) + s
}

File: lib/python/src/gencmu/_unicode.py
"""The character table every library shares, grammars/unicode.txt (engine §1)."""

from __future__ import annotations

from bisect import bisect_right

from ._errors import GencmuError

_SPACES = frozenset(
    [*range(0x09, 0x0E), 0x20, 0x85, 0xA0, 0x1680, *range(0x2000, 0x200B), 0x2028, 0x2029, 0x202F, 0x205F, 0x3000]
)


class UnicodeTable:
    """Character classes and simple lower-case mappings."""

    def __init__(self, text: str) -> None:
        marks: list[tuple[int, int]] = []
        alphas: list[tuple[int, int]] = []
        self.lower: dict[int, int] = {}
        self.version = ""
        for number, line in enumerate(text.splitlines(), 1):
            fields = line.split()
            if not fields:
                continue
            try:
                if fields[0] == "unicode":
                    self.version = fields[1]
                elif fields[0] == "mark":
                    marks.append((int(fields[1], 16), int(fields[2], 16)))
                elif fields[0] == "alpha":
                    alphas.append((int(fields[1], 16), int(fields[2], 16)))
                elif fields[0] == "lower":
                    self.lower[int(fields[1], 16)] = int(fields[2], 16)
                else:
                    raise ValueError(fields[0])
            except (ValueError, IndexError) as error:
                raise GencmuError(f"unreadable line: {line!r}", document="unicode.txt", line=number) from error
        marks.sort()
        alphas.sort()
        self._mark_starts = [start for start, _ in marks]
        self._mark_ends = [end for _, end in marks]
        self._alpha_starts = [start for start, _ in alphas]
        self._alpha_ends = [end for _, end in alphas]
        self._classes: dict[str, str] = {}

    @staticmethod
    def _within(code: int, starts: list[int], ends: list[int]) -> bool:
        index = bisect_right(starts, code) - 1
        return index >= 0 and code <= ends[index]

    def character_class(self, char: str) -> str:
        cached = self._classes.get(char)
        if cached is not None:
            return cached
        code = ord(char)
        if code in _SPACES:
            result = "space"
        elif 0x30 <= code <= 0x39:
            result = "digit"
        elif self._within(code, self._mark_starts, self._mark_ends):
            result = "mark"
        elif self._within(code, self._alpha_starts, self._alpha_ends):
            result = "alpha"
        else:
            result = "other"
        self._classes[char] = result
        return result

    def lowercase(self, text: str) -> str:
        lower = self.lower
        return "".join(chr(lower.get(ord(char), ord(char))) for char in text)
